lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs

```diff
@@ -161,8 +193,188 @@ mkNetworkLayer j = NetworkLayer
+    (lbhs', nbhs) <- fetchBackwards lbhs [] 0 t
+    -- The new unstable blocks is the current local blocks up to where they
+    -- meet the blocks fetched starting from the tip, with the fetched
+    -- blocks appended.
+    pure $!! appendUnstableBlocks k lbhs' nbhs
+  where
+    -- | Fetch blocks backwards starting from the given id. If fetched blocks
+    -- overlap the local blocks, the excess local blocks will be dropped.
+    fetchBackwards
+        :: UnstableBlocks
+        -- ^ Current local unstable blocks
+        -> [(Hash "BlockHeader", BlockHeader, Quantity "block" Natural)]
+        -- ^ Accumulator of fetched blocks
+        -> Natural
+        -- ^ Accumulator for number of blocks fetched
+        -> Hash "BlockHeader"
+        -- ^ Starting point for block fetch
+        -> m (UnstableBlocks, [(Hash "BlockHeader", BlockHeader, Quantity "block" Natural)])
+    fetchBackwards ubs ac len t = do
+        (tipHeader, tipHeight) <- getBlockHeader t
+        -- Push the remote block.
+        let ac' = ((t, tipHeader, tipHeight):ac)
+         -- Pop off any overlap.
+        let ubs' = dropStartingFromSlot tipHeader ubs
+        -- If remote blocks have met local blocks, or if more than k have been
+        -- fetched, or we are at the genesis, then stop.
+        -- Otherwise, continue from the parent of the current tip.
+        let intersected = unstableBlocksTipId ubs' == Just (prevBlockHash tipHeader)
+        let bufferFull = len + 1 >= k
+        let atGenesis = slotId tipHeader == SlotId 0 0
+        if intersected || bufferFull || atGenesis
+            then pure (ubs', ac')
+            else fetchBackwards ubs' ac' (len + 1) (prevBlockHash tipHeader)
+
+-- | The tip block header of the unstable blocks, if it exists.
+unstableBlocksTip :: UnstableBlocks -> Maybe BlockHeader
+unstableBlocksTip (UnstableBlocks Empty _) = Nothing
+unstableBlocksTip (UnstableBlocks (_ubs :|> (_, bh)) _) = Just bh
+
+-- | The tip block id of the unstable blocks, if it exists.
+unstableBlocksTipId :: UnstableBlocks -> Maybe (Hash "BlockHeader")
+unstableBlocksTipId (UnstableBlocks Empty _) = Nothing
+unstableBlocksTipId (UnstableBlocks (_ubs :|> (t, _)) _) = Just t
+
+-- | Add recently fetched block headers to the unstable blocks. This will drop
+-- the oldest block headers to ensure that there are at most /k/ items in the
+-- sequence.
+appendUnstableBlocks
+    :: Natural
+    -- ^ Maximum length of sequence.
+    -> UnstableBlocks
+    -- ^ Current unstable block headers, with rolled back blocks removed.
+    -> [(Hash "BlockHeader", BlockHeader, Quantity "block" Natural)]
+    -- ^ Newly fetched block headers to add.
+    -> UnstableBlocks
+appendUnstableBlocks k (UnstableBlocks ubs h) bs =
+    UnstableBlocks (ubs `appendBounded` more) h'
+  where
+    more = Seq.fromList [(a, b) | (a, b, _) <- bs]
+    -- New block height is the height of the new tip block.
+    h' = maybe h thd3 (lastMay bs)
```

Comment:
Slight taste preference:

```hs
h' = maybe h (view position @3) (lastMay bs)
```

Somewhat more readable as it doesn't require decrypting a function acronym :sweat_smile:  (which I didn't know existed)